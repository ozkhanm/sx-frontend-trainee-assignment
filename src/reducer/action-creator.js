const ActionType = {
    GET_ARTICLES: `GET_ARTICLES`,
    CHANGE_LOADING_STATUS: `CHANGE_LOADING_STATUS`
};

const ActionCreator = {
    getArticles: (articles) => ({
        type: ActionType.GET_ARTICLES,
        payload: articles
    }),
    changeLoadingStatus: (status) => ({
        type: ActionType.CHANGE_LOADING_STATUS,
        payload: status
    })
};

export {ActionCreator, ActionType};