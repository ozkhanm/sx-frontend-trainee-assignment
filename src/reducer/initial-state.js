const initialState = {
    articles: [],
    isDataLoaded: false
};

export default initialState;