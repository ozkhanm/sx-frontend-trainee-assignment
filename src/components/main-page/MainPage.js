import React, {useEffect} from "react";
import {Link} from "react-router-dom";
import {connect} from "react-redux";

import PageHeader from "../page-header/PageHeader";
import LoadingPage from "../loading-page/LoadingPage";

import {getDate} from "../../utils";
import {Operation} from "../../reducer/reducer";
import {ActionCreator} from "../../reducer/action-creator";

const MainPage = (props) => {
    const {articles, isDataLoaded, getArticles, changeRefreshStatus} = props;

    useEffect(() => {
        const refreshInterval = setInterval(() => {
            changeRefreshStatus(true);
            getArticles();
          }, 60000);

        return () => {
            clearInterval(refreshInterval);
        };
    });

    const getListArticles = (articles) => {
        return articles.map((it) => {
            if (it !== null) {
                const formattedDate = getDate(it.time);
                const additionalInfo = `by: ${it.by} | at: ${formattedDate}`;

                return (
                    <li key={it.id} className="item">                                    
                        <div className="content-block">
                            <Link to={`/${it.id}`} className="item-link">{it.title}</Link>
                            <div>
                                <p className="additional-content-block">{additionalInfo}</p>
                            </div>
                        </div>
                        <p className="score">{it.score}</p>
                    </li>
                );
            } else {
                return;
            }
        });
    };

    if (isDataLoaded) {
        return (
            <React.Fragment>
                <PageHeader page={`MAIN_PAGE`}/>
                <div className="container">
                    <ul className="list">
                        {getListArticles(articles)}
                    </ul>
                </div>
            </React.Fragment>
        );
    } else {
        return <LoadingPage/>;
    }
};

const mapStateToProps = (state) => ({
    articles: state.articles,
    isDataLoaded: state.isDataLoaded
});

const mapDispatchToProps = (dispatch) => ({
    getArticles() {
        dispatch(Operation.getArticles());
    },
    changeRefreshStatus(status) {
        dispatch(ActionCreator.changeRefreshStatus(status));
    }
})

export {MainPage};
export default connect(mapStateToProps, mapDispatchToProps)(MainPage);