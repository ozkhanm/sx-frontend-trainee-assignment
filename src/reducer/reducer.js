import initialState from "./initial-state";
import {ActionType, ActionCreator} from "./action-creator";

const Operation = {
    getArticles: () => (dispatch, getState, api) => {
        return api.get(`/newstories.json?print=pretty`)
            .then((response) => {
                const ids = response.data.slice(0, 100);
                
                return Promise.all(ids.map((it) => api.get(`item/${it}.json?print=pretty`)));
            })
            .then((response) => {
                const articles = response.slice().map((it) => it.data);

                dispatch(ActionCreator.getArticles(articles));
            });
    }
};

const reducer = (state = initialState, action) => {
    switch (action.type) {
        case ActionType.CHANGE_LOADING_STATUS:
            return Object.assign({}, state, {
                isDataLoaded: action.payload
            });

        case ActionType.GET_ARTICLES:
            return Object.assign({}, state, {
                articles: action.payload
            });
    }

    return state;
};

export {reducer, Operation};