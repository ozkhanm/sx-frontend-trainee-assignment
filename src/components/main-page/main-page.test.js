import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import {createStore} from "redux";
import {BrowserRouter} from 'react-router-dom';

import {MainPage} from "./MainPage";

import {reducer} from "../../reducer/reducer";

const articles= [
    {
        "by" : "yarapavan",
        "descendants" : 0,
        "id" : 30406481,
        "score" : 1,
        "time" : 1645371224,
        "title" : "How to Lose Time and Money",
        "type" : "story",
        "url" : "https://prachinain.medium.com/how-to-lose-time-and-money-e2d8c54c5c4d"
    },
    {
        "by" : "idletom",
        "descendants" : 0,
        "id" : 30406473,
        "score" : 1,
        "time" : 1645371191,
        "title" : "Technical Analysis of the DDoS Attacks Against Ukrainian Websites",
        "type" : "story",
        "url" : "https://www.cadosecurity.com/technical-analysis-of-the-ddos-attacks-against-ukrainian-websites/"
    },
    {
        "by" : "wrp",
        "descendants" : 1,
        "id" : 30406031,
        "kids" : [ 30406036 ],
        "score" : 2,
        "time" : 1645367930,
        "title" : "Absence of evidence is not evidence of absence: Bayesian view? (2021)",
        "type" : "story",
        "url" : "https://stats.stackexchange.com/questions/512678/absence-of-evidence-is-not-evidence-of-absence-what-does-bayesian-probability-h"
    }
];

it(`Main page renders correctly if data fetched`, () => {
    const tree = renderer
        .create(
            <Provider store={createStore(reducer)}>
                <BrowserRouter>
                    <MainPage articles={articles} isDataLoaded={true}/>
                </BrowserRouter>
            </Provider>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});
  
it(`Main page renders correctly if no data fetched`, () => {
    const tree = renderer
        .create(
            <Provider store={createStore(reducer)}>
                <BrowserRouter>
                    <MainPage articles={[]} isDataLoaded={false}/>
                </BrowserRouter>
            </Provider>)
        .toJSON();
  
    expect(tree).toMatchSnapshot();
});