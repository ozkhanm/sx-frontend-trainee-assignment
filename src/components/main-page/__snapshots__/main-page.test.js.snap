// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page renders correctly if data fetched 1`] = `
Array [
  <div>
    <div
      className="header-controls"
    >
      <a
        className="header-controls-button header-controls-refresh-button "
        onClick={[Function]}
      />
    </div>
    <p
      className="header"
    >
      Hacker News
    </p>
  </div>,
  <div
    className="container"
  >
    <ul
      className="list"
    >
      <li
        className="item"
      >
        <div
          className="content-block"
        >
          <a
            className="item-link"
            href="/30406481"
            onClick={[Function]}
          >
            How to Lose Time and Money
          </a>
          <div>
            <p
              className="additional-content-block"
            >
              by: yarapavan | at: 20.02.2022, 18:33
            </p>
          </div>
        </div>
        <p
          className="score"
        >
          1
        </p>
      </li>
      <li
        className="item"
      >
        <div
          className="content-block"
        >
          <a
            className="item-link"
            href="/30406473"
            onClick={[Function]}
          >
            Technical Analysis of the DDoS Attacks Against Ukrainian Websites
          </a>
          <div>
            <p
              className="additional-content-block"
            >
              by: idletom | at: 20.02.2022, 18:33
            </p>
          </div>
        </div>
        <p
          className="score"
        >
          1
        </p>
      </li>
      <li
        className="item"
      >
        <div
          className="content-block"
        >
          <a
            className="item-link"
            href="/30406031"
            onClick={[Function]}
          >
            Absence of evidence is not evidence of absence: Bayesian view? (2021)
          </a>
          <div>
            <p
              className="additional-content-block"
            >
              by: wrp | at: 20.02.2022, 17:38
            </p>
          </div>
        </div>
        <p
          className="score"
        >
          2
        </p>
      </li>
    </ul>
  </div>,
]
`;

exports[`Main page renders correctly if no data fetched 1`] = `
Array [
  <div>
    <div
      className="header-controls"
    >
      <a
        className="header-controls-button header-controls-refresh-button "
        onClick={[Function]}
      />
    </div>
    <p
      className="header"
    >
      Hacker News
    </p>
  </div>,
  <h1
    className="load"
  >
    Загрузка данных...
  </h1>,
]
`;
